contrib/dracut/90zfs/zfs-load-key.sh.in

```diff
@@ -1,55 +1,48 @@
 [ -f /lib/dracut-lib.sh ] && dracutlib=/lib/dracut-lib.sh
 [ -f /usr/lib/dracut/modules.d/99base/dracut-lib.sh ] && dracutlib=/usr/lib/dracut/modules.d/99base/dracut-lib.sh
+# shellcheck source=./lib-zfs.sh.in
 . "$dracutlib"
 
 # load the kernel command line vars
-[ -z "$root" ] && root=$(getarg root=)
+[ -z "$root" ] && root="$(getarg root=)"
 # If root is not ZFS= or zfs: or rootfstype is not zfs then we are not supposed to handle it.
-[ "${root##zfs:}" = "${root}" -a "${root##ZFS=}" = "${root}" -a "$rootfstype" != "zfs" ] && exit 0
+[ "${root##zfs:}" = "${root}" ] && [ "${root##ZFS=}" = "${root}" ] && [ "$rootfstype" != "zfs" ] && exit 0
 
 # There is a race between the zpool import and the pre-mount hooks, so we wait for a pool to be imported
 while true; do
     zpool list -H | grep -q -v '^$' && break
-    [[ $(systemctl is-failed zfs-import-cache.service) == 'failed' ]] && exit 1
-    [[ $(systemctl is-failed zfs-import-scan.service) == 'failed' ]] && exit 1
+    [ "$(systemctl is-failed zfs-import-cache.service)" = 'failed' ] && exit 1
+    [ "$(systemctl is-failed zfs-import-scan.service)" = 'failed' ] && exit 1
     sleep 0.1s
 done
 
 # run this after import as zfs-import-cache/scan service is confirmed good
-if [[ "${root}" = "zfs:AUTO" ]] ; then
-    root=$(zpool list -H -o bootfs | awk '$1 != "-" {print; exit}')
+if [ "${root}" = "zfs:AUTO" ] ; then
+    root="$(zpool list -H -o bootfs | awk '$1 != "-" {print; exit}')"
 else
     root="${root##zfs:}"
     root="${root##ZFS=}"
 fi
 
 # if pool encryption is active and the zfs command understands '-o encryption'
-if [[ $(zpool list -H -o feature@encryption $(echo "${root}" | awk -F\/ '{print $1}')) == 'active' ]]; then
-    # check if root dataset has encryption enabled
-    if $(zfs list -H -o encryption "${root}" | grep -q -v off); then
-        # figure out where the root dataset has its key, the keylocation should not be none
-        while true; do
-            if [[ $(zfs list -H -o keylocation "${root}") == 'none' ]]; then
-                root=$(echo -n "${root}" | awk 'BEGIN{FS=OFS="/"}{NF--; print}')
-                [[ "${root}" == '' ]] && exit 1
-            else
-                break
-            fi
-        done
+if [ "$(zpool list -H -o feature@encryption $(echo "${root}" | awk -F\/ '{print $1}'))" = 'active' ]; then
+    # if the root dataset has encryption enabled
+    ENCRYPTIONROOT=$(zfs get -H -o value encryptionroot "${ZFS_DATASET}")
+    if ! [ "${ENCRYPTIONROOT}" = "-" ]; then
         # decrypt them
         TRY_COUNT=5
-        while [ $TRY_COUNT != 0 ]; do
-            zfs load-key "$root" <<< $(systemd-ask-password "Encrypted ZFS password for ${root}: ")
-            [[ $? == 0 ]] && break
-            ((TRY_COUNT-=1))
+        while [ $TRY_COUNT -gt 0 ]; do
+            printf "%s" "$(systemd-ask-password "Encrypted ZFS password for ${root}: ")" | zfs load-key "${ENCRYPTIONROOT}" && break
+            TRY_COUNT-=1
         done
     fi
 fi
+
```

Comment:
this needs a newline but `mount-zfs.sh.in` doesn't?

Reply:
it was not added out of necessity, just an artifact of the editor.